gateway-server/src/main/java/org/apache/knox/gateway/util/TokenMigrationTool.java

```diff
@@ -0,0 +1,232 @@
+import java.util.HashSet;
+import java.util.Locale;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.atomic.AtomicInteger;
+
+import org.apache.knox.gateway.i18n.messages.MessagesFactory;
+import org.apache.knox.gateway.services.security.AliasService;
+import org.apache.knox.gateway.services.security.AliasServiceException;
+import org.apache.knox.gateway.services.security.token.TokenMetadata;
+import org.apache.knox.gateway.services.security.token.TokenStateService;
+import org.apache.knox.gateway.services.token.impl.TokenStateServiceMessages;
+
+public class TokenMigrationTool {
+
+  private static final String TOKEN_ALIAS_SUFFIX_DELIM = "--";
+  private static final String TOKEN_ISSUE_TIME_POSTFIX = TOKEN_ALIAS_SUFFIX_DELIM + "iss";
+  private static final String TOKEN_MAX_LIFETIME_POSTFIX = TOKEN_ALIAS_SUFFIX_DELIM + "max";
+  private static final String TOKEN_META_POSTFIX = TOKEN_ALIAS_SUFFIX_DELIM + "meta";
+  private static final TokenStateServiceMessages LOG = MessagesFactory.get(TokenStateServiceMessages.class);
+
+  private final AliasService aliasService;
+  private final TokenStateService tokenStateService;
+  private final PrintStream out;
+
+  private int progressCount = 10;
+  private boolean archiveMigratedTokens;
+  private boolean migrateExpiredTokens;
+  private boolean verbose;
+
+  public TokenMigrationTool(AliasService aliasService, TokenStateService tokenStateService, PrintStream out) {
+    this.aliasService = aliasService;
+    this.tokenStateService = tokenStateService;
+    this.out = out;
+  }
+
+  public void setProgressCount(int progressCount) {
+    this.progressCount = progressCount;
+  }
+
+  public void setArchiveMigratedTokens(boolean archiveMigratedTokens) {
+    this.archiveMigratedTokens = archiveMigratedTokens;
+  }
+
+  public void setMigrateExpiredTokens(boolean migrateExpiredTokens) {
+    this.migrateExpiredTokens = migrateExpiredTokens;
+  }
+
+  public void setVerbose(boolean verbose) {
+    this.verbose = verbose;
+  }
+
+  public void migrateTokensFromGatewayCredentialStore() {
+    try {
+      final Map<String, TokenData> tokenDataMap = new ConcurrentHashMap<>();
+      final long start = System.currentTimeMillis();
+      String logMessage = "Loading token aliases from the __gateway credential store. This could take a while.";
+      log(logMessage);
+      final Map<String, char[]> passwordAliasMap = aliasService.getPasswordsForGateway();
+      log("Token aliases loaded in " + (System.currentTimeMillis() - start) + " milliseconds");
+      String alias;
+      for (Map.Entry<String, char[]> passwordAliasMapEntry : passwordAliasMap.entrySet()) {
+        alias = passwordAliasMapEntry.getKey();
+        processAlias(passwordAliasMap, passwordAliasMapEntry, alias, tokenDataMap);
+      }
+
+      final long migrationStart = System.currentTimeMillis();
+      final AtomicInteger count = new AtomicInteger(0);
```

Comment:
Is there a reason why we use AtomicInteger + ConcurrentHashMap here? Is this method called from multiple threads?

Reply:
No, there is no multi-threading here. I needed to use AtomicInteger because of the stream below accepts a final `count` variable.